core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1233,6 +1233,22 @@ abstract class RDD[T: ClassTag](
           (i, iter) => iter.map((i % curNumPartitions, _))
```

Comment:
I think this option is fine:

```scala
  def treeAggregate[U](
    zeroValue: U,
    seqOp: JFunction2[U, T, U],
    combOp: JFunction2[U, U, U],
    depth: Int,
    finalAggregateOnExecutor: Boolean): U = {
...
```

To use `finalAggregateOnExecutor`, all parameters have to be fully specified. In this way, none of things will be broken.

Reply:
i don't think we can do that as we did not want to cause binary incompabilities. that was the whole reason i had to revert back to the config approach. refer to this thread: https://github.com/apache/spark/pull/33644#discussion_r687122613

Reply:
Overloading also does not work due to the reasons given in the above thread

Reply:
Does this signature alone requires the changes in MiMa exclusion? This is a completely new separate signature that shouldn't affect other signatures apparently. I think the exclusion was added for other signatures you added like:

```
  def treeAggregate[U](
    zeroValue: U,
    seqOp: JFunction2[U, T, U],
    combOp: JFunction2[U, U, U],
    depth: Int): U = {
    ...
```
?

Reply:
yes, we can add a new method with signature that looks like this: 
```
  def treeAggregate[U](
    zeroValue: U,
    seqOp: JFunction2[U, T, U],
    combOp: JFunction2[U, U, U],
    depth: Int,
    finalAggregateOnExecutor: Boolean): U = {
...
```
you're right, we dont need to worry about Mima excludes then. 
but that is not the problem. the problem is that this method will need to be overloaded by the other treeAggregate methods. for example, the other treeAggregate methods will have to call this version of treeAggregate. this is an issue because of zeroValue:U is only inspected by the compiler to resolve the overloaded methods. Due to that the compiler throws ambiguous reference to overloaded definition, and this cannot be resolved..

there are only 2 workarounds: 
1. create a new method with a different name but same signature with the new parameter. something like `treeAggregateExecutor` and have somewhat duplicate code in that method
2. use config  approach

Reply:
Okie, got the problem now. Configuration SGTM

Reply:
Actually, I like @HyukjinKwon's proposal.
Add a new `treeAggregate` method to RDD which has all the parameters explicitly specified (no defaults, no currying).
And have everything else delegate to it.

This should take care of @srowen's concerns about binary/source compatibility (existing methods remain as is), while introducing a new method in scala api which allows for per method customization (and not global config affecting all aggregates).

Thoughts ?

In RDD:
```
  def treeAggregate[U: ClassTag](zeroValue: U)(
      seqOp: (U, T) => U,
      combOp: (U, U) => U,
      depth: Int = 2): U = withScope {
    treeAggregate(zeroValue, seqOp, combOp, depth, false)
  }

  def treeAggregate[U: ClassTag](
      zeroValue: U, 
      seqOp: (U, T) => U,
      combOp: (U, U) => U,
      depth: Int,
      finalAggregateOnExecutor: Boolean): U = withScope {

  // modified method taking finalAggregateOnExecutor into account

  }
```

java api to mirror and delegate to the scala api as appropriate.

Thoughts @HyukjinKwon, @srowen, @akpatnam25, @venkata91 ?

Reply:
i just made the change. i think i did not read @HyukjinKwon's suggestion properly earlier, as I did not realize that he was proposing a method that was not curried. i think this approach is the best way, and also does not break binary/source incompatibility.